integration/watch.ts

```diff
@@ -1,86 +1,150 @@
+      return;
+    }
 
-    chokidar
-      .watch("*/*.bin", { ignoreInitial: true, usePolling })
-      .on('all', yarnRunHandler(yarn, 'bin2ts'));
-}
+    triggerPath = triggerPath.replace(/\\/g, '/'); // windows
+    const relativePath = triggerPath.replace(/^integration\//, '');
 
-const colors = {
-    none: '',
-    reset: '\x1b[0m',
-    green: '\x1b[32m',
-    red: '\x1b[31m',
-    yellow: '\x1b[33m',
-    blue: '\x1b[34m',
-    magenta: '\x1b[35m',
-    cyan: '\x1b[36m'
+    yarnRun(yarn, task, triggerPath, relativePath);
+  };
 }
 
-function yarnRunHandler(yarn: string, task: string) {
-    return (event: "add" | "addDir" | "change" | "unlink" | "unlinkDir", path: string) => {
-        if (event !== 'add' && event !== 'change') {
-            return;
-        }
+function yarnRun(yarn: string, task: string, header: string, taskArgument?: string) {
+  const yarnArgs = taskArgument ? [task, taskArgument] : [task];
 
-        path = path.replace(/\\/g, "/"); // windows
+  console.log(formatLog(colors.green, header, task, `${yarn} ${yarnArgs.join(' ')}`));
 
-        yarnRun(yarn, task, path);
+  const yarnProcess = spawn(yarn, yarnArgs);
+  yarnProcess.stdout.on('data', (data: Buffer) => console.log(formatLog(colors.none, header, task, data.toString())));
+  yarnProcess.stderr.on('data', (data: Buffer) => console.error(formatLog(colors.red, header, task, data.toString())));
+  yarnProcess.on('error', (err: Error) => console.error(formatLog(colors.red, header, task, err.message)));
+  yarnProcess.on('close', (code: number) => {
+    if (code !== 0) {
+      console.error(formatLog(colors.red, header, task, `Exited with code ${code}`));
+    } else {
+      console.log(formatLog(colors.green, header, task, 'Done'));
     }
+  });
 }
 
-function yarnRun(yarn: string, task: string, path: string) {
-    const yarnArgs = [task, path];
-
-    console.log(formatLog(colors.green, task, path, `${yarn} ${yarnArgs.join(' ')}`));
+function sourceHandler(yarn: string, task: string, tests: string[]) {
+  return async (event: FsEvent, triggerPath: string) => {
+    if (event !== 'change') return;
 
-    const yarnProcess = spawn(yarn, yarnArgs);
-    yarnProcess.stdout.on('data', (data: Buffer) => console.log(formatLog(colors.none, task, path, data.toString())));
-    yarnProcess.stderr.on('data', (data: Buffer) => console.error(formatLog(colors.red, task, path, data.toString())));
-    yarnProcess.on('error', (err: Error) => console.error(formatLog(colors.red, task, path, err.message)));
-    yarnProcess.on('close', (code: number) => {
-        if (code !== 0) {
-            console.error(formatLog(colors.red, task, path, `Exited with code ${code}`));
-        }
-    });
+    triggerPath = triggerPath.replace(/\\/g, '/');
+    if (tests.length === 0) {
+      let notice = `Source changed! Press [enter] to recompile all .proto files or rerun this command as 'yarn watch [TEST, ...]'. See 'yarn watch --help'.`;
```

Comment:
Maybe `src/ changed! ... to recodegen / to regenerate`...

Reply:
The message is prefixed with the affected path: `src/main.ts [watch] Source changed! ...`.

I used "source" to indicate that it's not the tests that have changed, but the actual plugin. Not so much to show which directory it is. I wonder if another name would be better to express this? "Plugin changed", "Code generation logic changed", "Source code changed!"?

Reply:
Changed to

`src/main.ts [watch] Plugin modified! Press [enter] to regenerate all integration tests or run 'yarn watch [TEST, ...]'. See 'yarn watch --help'.`

for now

Reply:
Ah gotcha, I hadn't actually tried it, so didn't realize `src/` was already part of the path. But that's great, thanks!